File: 205.同构字符串.py
#
# @lc app=leetcode.cn id=205 lang=python
#
# [205] 同构字符串
#

# @lc code=start
class Solution(object):
    def isIsomorphic(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        if len(s) != len(t):
            return False
        d = {}
        usedChar = set()
        for a, b in zip(s, t):
            if a not in d:
                if b not in usedChar:
                    usedChar.add(b)
                    d[a] = b
                else:
                    return False
            else:
                if d[a] != b:
                    return False
        return True
        
# @lc code=end

File: 205.同构字符串.cpp
/*
 * @lc app=leetcode.cn id=205 lang=cpp
 *
 * [205] 同构字符串
 */

// @lc code=start
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if(s.size() != t.size()){
            return false;
        }
        unordered_map<char, char>maps;
        unordered_map<char, char>mapt;
        for(int i = 0; i < s.size(); i++){
            if(!maps[s[i]])
                maps[s[i]] = t[i];
            else if(maps[s[i]] != t[i])
                return false;
            if(!mapt[t[i]])
                mapt[t[i]] = s[i];
            else if(mapt[t[i]] != s[i])
                return false;
        }
        return true;
    }
};
// @lc code=end

